### ISSUE TITLE
Terminated slash in route prefixes lead to not found

### ISSUE THREAD

**Issue by StarpTech** [fastify/fastify#584]
Request to `http://localhost:3000/test/foo` will return 404 because of the terminated slash `/test/`. The correct URL is `localhost:3000/test//foo`. In my opinion, for the best dev experience, we should unify this and don't allow a terminated `/` in prefix as well as no `/` in routes.

Correct:
```
route: `/route`
prefix: `/foo`
```

Wrong:
```
route: `route`
prefix: `/foo/`
```

```js
'use strict'

const fastify = require('../fastify')()

const schema = {
  schema: {
    response: {
      200: {
        type: 'object',
        properties: {
          hello: {
            type: 'string'
          }
        }
      }
    }
  }
}

fastify.register(
  (f,o, next)=>{
      f.get('/foo', function (request, reply) {
        reply.send({ hello: 'world' })
      })
      next()
  },
  { prefix: '/test/' }
)

fastify
  .get('/', schema, function (req, reply) {
    reply
      .send({ hello: 'world' })
  })

fastify.listen(3000, err => {
  if (err) throw err
  console.log(`server listening on ${fastify.server.address().port}`)
})
```

**Comment by delvedor** [comment]
There are different solution for this "problem", spoiler alert: "doing nothing" is my solution.

1. Check every prefix and if there is a final `/` remove it
 👍  solve the probem
 👎  if a user declares a route as `foo` it will not work (and the router will throw an error because the first char **must** be `/`
2. write some unnecessarily complicated logic internally to handle the problem.
3. doing nothing, add a `/` at the end of the route does not have any sense, the user should not do it .

**Comment by StarpTech** [comment]
mmh then I found a bug. Registering a route inside a plugin without a `/` don't lead to an error. This inconsistency is what I mean.

```js
fastify.register(
  (f,o, next)=>{
      f.get('foo', function (request, reply) {
        reply.send({ hello: 'world' })
      })
      next()
  },
  { prefix: '/test' }
)

fastify
  .get('/foo', schema, function (req, reply) {
    reply
      .send({ hello: 'world' })
  })
```

**Comment by jsumners** [comment]
This is complicated. Essentially, you have the following situations to resolve:

1. `http://example.com/foo/bar` -- this should be an absolute URI to a resource
1. `http://example.com/foo/bar/` -- this should be an absolute URI to a collection of resources
1. `http://example.com/foo//bar/` -- this should be invalid because _each `/` matters_
1. `http://example.ccom//foo/bar/` -- invalid because it should translate to `http://foo/bar`

See https://tools.ietf.org/html/rfc3986#section-5.4

But what are people expecting?

+ Are they expecting `/foo/bar/` to be equivalent to `/foo/bar`? Hard to tell and should necessitate a configuration parameter with a sane default (`false`).
+ Are they expecting `/foo//bar/` to be equivalent to `/foo/bar/`? I think so.
+ Are they expecting `//foo/bar/` to be equivalent to `/foo/bar/`? I definitely think so.

**Comment by jsumners** [comment]
Also, while trying to find a definitive answer on trailing slashes I ran across this ancient article that may or may not still apply -- https://cdivilly.wordpress.com/2014/03/11/why-trailing-slashes-on-uris-are-important/

**Comment by StarpTech** [comment]
From my experience, I would say that any kind of route has to be prefixed with a `/` everything should throw an error. @jsumners do you see an error in this?

**Comment by jsumners** [comment]
@StarpTech I do not understand your statement. It seems to be missing a word or two.

**Comment by StarpTech** [comment]
@jsumners try it again :D

**Comment by jsumners** [comment]
Notice that I specifically called out that scenario. It's _impossible_ to make that assumption. It _must_ be left up to the application author to make that determination.

**Comment by delvedor** [comment]
@StarpTech if yuou declare a route and the first char is not `/` or `*` [`find-my-way`](https://github.com/delvedor/find-my-way) will throw an error.
We add the `/`at the beginning of the prefixes (if it is missing) because of https://github.com/fastify/fastify/pull/132.

**Comment by StarpTech** [comment]
@delvedor the issue was about to support prefixes without to type a `/` but why it's the same for declared routes?

**Comment by delvedor** [comment]
The route declaration is handled by find-my-way.
Checkout https://github.com/fastify/fastify/issues/130 to understand why that was needed.

**Comment by StarpTech** [comment]
@delvedor  It's fine for prefixes 

```diff
fastify.register(
  (f,o, next)=>{
      f.get('foo', function (request, reply) {
        reply.send({ hello: 'world' })
      })
      next()
  },
++  { prefix: 'test' }
)
```
but why here
```diff
fastify.register(
  (f,o, next)=>{
++      f.get('foo', function (request, reply) {
        reply.send({ hello: 'world' })
      })
      next()
  },
  { prefix: 'test' }
)
```

**Comment by delvedor** [comment]
Because we chain the prefix and the route path before save the route in `find-my-way`.
https://github.com/fastify/fastify/blob/5ac50d37550ee0e5218e47d19fa724299da0ed71/fastify.js#L407-L409

**Comment by StarpTech** [comment]
You don't think it's consistent to throw this `f.get('foo')` but don't this `f.get('foo')` inside a prefix context? In my opinion, it would be clearer that any route has start with `/` doesn't matter where you define it.

**Comment by jsumners** [comment]
Issue #598 would be solved by:

> Are they expecting /foo/bar/ to be equivalent to /foo/bar? Hard to tell and should necessitate a configuration parameter with a sane default (false).

**Comment by delvedor** [comment]
IMHO ´/test/´ **is not** equal to `/test`, since every `/` means a new nested route.
If you use express, `/test/` and `/test` are treated as equals, while if you use hapi they are not.

@fastify/fastify thoughts?

**Comment by jsumners** [comment]
I think it should be configurable.

**Comment by delvedor** [comment]
@jsumners that could be nice, but I'm pretty sure it will compromise the performances.

**Comment by jsumners** [comment]
I think that would depend on when the configuration is applied. Routes are established at bootup. So applying the configuration at bootup should be equal.

**Comment by delvedor** [comment]
Yes and no, you will pass the configuration at initialization time, but do not forget that the routing algorithm is handled by [`find-my-way`](https://github.com/delvedor/find-my-way) which is a radix tree.
Add this kind of configuration means change the search algorithm, and checking at every loop that configuration.

**Comment by jsumners** [comment]
```js
function addRoute (path, handler) {
  if (config.trimTrailingSlash && path.slice(-1) === '/') {
    register(path, handler)
    register(path.slice(0, -1), handler)
  } else {
    register(path, handler)
  }
}
```

**Comment by delvedor** [comment]
Nice catch @jsumners! :)
@fastify/fastify what's your thoughts about this?

**Comment by StarpTech** [comment]
LGTM, but what's with the circumstance above? Why throws route `foo` an error but `foo` inside a prefix not? Is it wrong to say that all paths have to start with `/` but ending `/` is configurable?

**Comment by StarpTech** [comment]
@delvedor could you answer my last question?

**Comment by delvedor** [comment]
@StarpTech sorry for the delay.
I quote myself:

> IMHO `/test/` is not equal to `/test`, since every `/` means a new nested route.
If you use express, `/test/` and `/test` are treated as equals, while if you use hapi they are not.

I have an opinion about this, but it looks like there is no common rule.
If we want to support both approaches we can do it (see @jsumners [comment](https://github.com/fastify/fastify/issues/584#issuecomment-355134797)).

**Comment by nwoltman** [comment]
It looks like there are really 2 issues here.

@jsumners Has solved the issue of how to configure routes that will match requests both with and without a trailing slash.

However, @StarpTech's issue about how route paths should be concatenated with the `prefix` option and whether or not paths should always start with a `/` has not been decided on yet.

My opinion on that is to mostly do nothing, but if a user is about to make a route with double slashes (like `/prefix//path`), we should normalize the path to remove the extra slash.

**Comment by delvedor** [comment]
> My opinion on that is to mostly do nothing, but if a user is about to make a route with double slashes (like /prefix//path), we should normalize the path to remove the extra slash.

Agree with this, someone wants to send a pr? :)

**Comment by jsumners** [comment]
Are we leaving this open until someone adds the configuration item for trailing slashes or has this been solved by #676?

**Comment by nwoltman** [comment]
I think #676 should be enough to close this. The trailing slashes configuration should probably be implemented in `find-my-way`.
